Archive/Imo/Imo1982Q1.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2024 Alex Brodbelt. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Brodbelt
+-/
+
+import Mathlib.Tactic.NormNum
+import Mathlib.Data.PNat.Basic
+
+/-!
+# IMO 1982 Q1
+
+The function $f(n)$ is defined for all postive integers $n$ and takes on
+non-negative integer values. Also, for all m, n
+
+$f (m + n) - f(m) - f(n) = 0 \text{ or } 1$
+$f(2) = 0, f(3) > 0, and f(9999) = 3333.$
+
+Determine f(1982).
+
+The solution is based on the one at the
+[Art of Problem Solving](https://artofproblemsolving.com/wiki/index.php/1982_IMO_Problems/Problem_1)
+website.
+-/
+
+namespace Imo1982Q1
+
+structure IsGood (f : ℕ+ → ℕ) : Prop where
+  /-- Satisfies the functional relation-/
+  rel: ∀ m n : ℕ+, f (m + n) = f m + f n ∨ f (m + n) = f m + f n + 1
+  f₂ : f 2 = 0
+  hf₃ : f 3 > 0
+  f_9999 : f 9999 = 3333
+
+namespace IsGood
+
+variable {f : ℕ+ → ℕ} (hf : IsGood f)
+include hf
+
+lemma f₁ : f 1 = 0 := by
+  have h : f 2 = 2 * f 1 ∨ f 2 = 2 * f 1 + 1 := by rw [two_mul]; apply hf.rel 1 1
+  rw [hf.f₂] at h
+  by_contra hf
+  rw [← ne_eq] at hf
+  have f1_pos : 0 < f 1 := by rw [lt_iff_le_and_ne]; exact ⟨zero_le (f 1), hf.symm⟩
+  rcases h with ( h₁ | h₂ )
+  · have : 2 * f 1 > 0 := by apply mul_pos (by norm_num) f1_pos
+    rw [← h₁] at this
+    exact lt_irrefl 0 this
+  · have : 2 * f 1 + 1 > 0 := by apply add_pos (mul_pos (by norm_num) f1_pos) (by norm_num)
+    rw [← h₂] at this
+    exact lt_irrefl 0 this
```

Comment:
Here's a shorter proof :)
```suggestion
  have h : f 2 = 2 * f 1 ∨ f 2 = 2 * f 1 + 1 := by rw [two_mul]; exact hf.rel 1 1
  obtain h₁ | h₂ := hf.f₂ ▸ h
  · rw [eq_comm, mul_eq_zero] at h₁
    apply h₁.resolve_left
    norm_num
  · cases Nat.succ_ne_zero _ h₂.symm
```
This adds the `Mathlib.Algebra.GroupWithZero.Defs` import. I admit the trick with `▸` is a bit cheeky; you can replace it with a `rw` if you want.

Reply:
Oh nice - what does `▸` do? Is this how you 'perform' rw in a proof term? I have been wondering how to perform what I would do in a rewrite within a proof term.

Reply:
The triangle `▸` is a macro for `Eq.subst`. Given `h : x = y`, `h ▸ e` will find instances of `x` in the type of `e` and replace them by `y`.

You can sometimes use it to golf proofs like this, but really, using `rw` is both more readable and robust.